### ISSUE TITLE
Ability to clear out old branches

### ISSUE THREAD

**Issue by Almenon** [eamodio/vscode-gitlens#1155]
At work I have waaayyyy too many old branches locally, the majority of which can be deleted. It would be cool if gitlens offered a command to delete all the branches that have been merged into the target branch (ex: main/dev) of the remote.

Command picker -> "Delete old branches that have already been merged into X" -> asks for target branch -> window opens up with list of old branches to be removed, asks if you are sure -> confirm

Stretch goal: Also have the ability to remove branches in the list in case there is one you want to keep.

**Comment by eamodio** [comment]
While there is no auto-suggest for which branches to delete, you can delete in bulk using the _Git Command Palette_ -> branch -> delete option. There you can multiselect all the branches to delete.

Would have to think more about the UX for integrating some option like that into the existing menu. Although this isn't something that I will likely add myself, but if the community wants to champion it and provide a PR (after discussions) I'm certainly open to it -- of course, depending on the complexity and scope of the changes required.

**Comment by flying-sheep** [comment]
There’s also an UI problem related to this:

1. I create a branch `old-branch`, push it, and using GitHub’s UI, merge it into master and delete it in the remote repo
2. I `git fetch origin --prune` to get rid of the stale remote-tracking branches. `git branch -a` shows that `remotes/origin/old-branch` is gone
2. The Branches tree view shows “old-branch ⇄ origin/old-branch” as if the remote-tracking branch still existed

I have two suggestions:

1. Make `fetch --prune` and `pull --prune` easily available for people. JetBrains IDEs just do it by default and I never regretted it (as long one uses GitHub and it allows to “restore branch”, nobody will)
2. The UI should only show “⇄ origin/*” if that remote-tracking branch still exists

That way people can just see at a glance which branches are not on the remote. And since that’s only branches you didn’t push yet or which you deleted on the remote, it allows a more informed decision about which ones can be deleted.

**Comment by krushndayshmookh** [comment]
> There’s also an UI problem related to this:
> 
> 1. I create a branch `old-branch`, push it (setting its upstream to the remote branch), and using GitHub’s UI, merge it into master and delete it in the remote repo
> 2. I `git fetch origin --prune` to get rid of the stale remote-tracking branches. `git branch -a` shows that `remotes/origin/old-branch` is gone
> 3. The Branches tree view shows “old-branch ⇄ origin/old-branch” as if the remote-tracking branch still existed
> 
> I have two suggestions:
> 
> 1. Make `fetch --prune` and `pull --prune` easily available for people. JetBrains IDEs just do it by default and I never regretted it (as long one uses GitHub and it allows to “restore branch”, nobody will)
> 2. The UI should only show “⇄ origin/*” if that remote-tracking branch still exists. I assume you show it because `.git/config` still has info about the upstream, even though it no longer exists
> 
> That way people can just see at a glance which branches are not on the remote. And since that’s only branches you didn’t push yet or which you deleted on the remote, it allows a more informed decision about which ones can be deleted.

Thanks, Man!

**Comment by flying-sheep** [comment]
OK, so both suggestions are implemented now:

1. `git.pruneOnFetch` and `git-graph.repository.fetchAndPrune` make this easier
2. there’s a “⚠” and red font for branches gone from the remote

Now the only thing missing is an easier way to delete those branches.

Maybe there could be an “×” button for branches, and for branches that have the “⚠ missing upstream” state, there will be no confirmation box, they’ll just be force-deleted?

**Comment by Pearce-Ropion** [comment]
@flying-sheep Did you ever figure out an easy way to delete the branches marked with a “⚠”?

**Comment by flying-sheep** [comment]
Yes, I have the [GitHub Pull Requests](https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-pull-request-github) extension installed and use the `pr.deleteLocalBranchesNRemotes` action (<kbd>Ctrl</kbd><kbd>Shift</kbd><kbd>P</kbd> then start typing “Delete local branches and remote repositories”)

It only works with GitHub that way of course, so this feature request would still be nice to see implemented.
